Add 'ascending' and 'descending' sorting options.

anoldguy opened the pull request:
This is an attempt to handle #5. 

I'm very new to rust, so I know there's some... not-idiomatic rust in here, and I'm happy to get some feedback on how it can be made better, or if I've missed something.

This let's you specify how you want the sorting handled,  either random (the default), ascending, or descending:

```toml
[default]
path = "/path/to/papers"
sorting = "ascending"

[eDP-1]
path = "/path/to/other/papers"
duration = "1m"
sorting = "descending"

[DP-2]
path = "/path/to/even/more"
duration = "5m"
```

That configuration would use a different sort for each monitor, a missing `sorting` value defaults to `random`.

Edit: Updated PR message to use the new sorting names.

Hoping to get some good feedback to make this PR better!

willglynn commented on daemon/src/wallpaper_info.rs:
```diff
@@ -12,6 +12,8 @@ pub struct WallpaperInfo {
     pub duration: Option<Duration>,
     #[serde(rename = "apply-shadow")]
     pub apply_shadow: Option<bool>,
+    #[serde(default = "set_sorting_default", deserialize_with = "order_deserializer")]
+    pub sorting: Option<String>
```
If this were an `enum Sorting` which implemented `Deserialize`, `serde` could deserialize it directly:

```rust
    #[serde(default = "set_sorting_default")]
    pub sorting: Option<Sorting>,
```

If `Sorting` additionally implemented [`Default`](https://doc.rust-lang.org/std/default/trait.Default.html), you can drop the `Option<T>` and tell `serde` to call `Sorting::default()` if it's not specified:

```rust
    #[serde(default)]
    pub sorting: Sorting,
```

You can `#[derive]` both `Deserialize` and `Default` along with all the other traits which might be useful:

```rust
#[derive(Debug, Copy, Clone, Default, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename = "kebab-case")]
pub enum Sorting {
    #[default]
    Random,
    Natural,
    Reverse,
}
```

anoldguy replied:
Fantastic! I was curious how to do that. String comparisons are my least favorite. I've absolutely done this, almost verbatim. :+1:

willglynn commented on daemon/src/surface.rs:
```diff
@@ -207,12 +209,35 @@ impl Surface {
+                    self.idx = 0
+                };
+
+                // This is a bit unweildy. We're sorting every time thru this loop. But since we're
+                // walking the directory each time, I guess it's not that bad. Sorting can't be slower
+                // than reading paths off the disk, right?
+                let img_path = if self.wallpaper_info.sorting.eq(&Some(String::from("natural"))) {
+                    // You can't sort an immutable Vec, so we have to clone it as mutable.
+                    let mut sorted: Vec<PathBuf> = files.clone();
+                    sorted.sort();
+                    sorted[self.idx].clone()
+                } else if self.wallpaper_info.sorting.eq(&Some(String::from("reverse"))) {
+                    // You can't sort an immutable Vec, so we have to clone it as mutable.
+                    let mut reversed: Vec<PathBuf> = files.clone();
+                    reversed.sort_by(|a,b| b.cmp(a));
+                    reversed[self.idx].clone()
+                } else {
+                    files[rand::random::<usize>() % files.len()].clone()
+                };
```
If you go with an `enum`, you'll probably want to switch over to:

```rust
let img_path = match self.wallpaper_info.sorting {
    Sorting::Random => { /* … */ },
    Sorting::Natural => { /* … */ },
    Sorting::Reverse => { /* … */ },
};
```

I note also that `files` isn't used after this block, which means you can move `files` rather than cloning it:

```rust
let mut sorted = files;
sorted.sort();
```

Likewise, since you need a `PathBuf` and the `Vec<PathBuf>` isn't used after picking one, you can move the selected `PathBuf` out of the `Vec` rather than cloning your selection:

```rust
let img_path = match sorting {
  /* … */
  Sorting::Random => {
    let index = rand::random::<usize>() % files.len();
    files.into_iter().nth(index).next().unwrap()
  }
};
```

anoldguy replied:
Again, yes, yes, yes! 

Learning more about copying and borrowing vs cloning things. Thanks for the tip!

willglynn reviewed (commented):
Good work! This is not my project, but I saw [your request for Rust feedback](https://hachyderm.io/@anoldguy/110599359717202845) and I can offer eyeballs 😄

danyspin97 commented:
Hello @anoldguy and @willglynn and thank you for your contribution!

@anoldguy  I am actually amazed to see that this is your first contribution, as working through arrays, enums and serde is usually advanced stuff, kudos to you :)

I think that having these 3 orders will be enough for most users. I may have spent too much time working with tech people so now `Natural` sound less natural to me compared to `ascending` and `descending` (sorry for the pun😂). I am not sure if it would be better to have `random`, `ascending` and `descending` instead to be more clear with non native english speakers. But that's just a small detail that we can decide before merging.

Although I like the index approach that you are using, one problem that might occur is that files are added or removed from the directory; in this case the index would point to a different image than the one it should be pointing to.

One other way would be to  use `current_img`: we search through the sorted array for `current_img` (getting the index), and then we take the next for `Natural` and the previous for `Reverse`. By using [`binary_search`](https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search), this operation should be fast enough to do it every time. What do you think of this solution?

danyspin97 commented on daemon/src/surface.rs:
```diff
@@ -207,12 +210,38 @@ impl Surface {
+                      }
+                    Sorting::Natural => {
+                        let mut sorted = files;
+                        sorted.sort();
```
Instead of calling `sort` here, we can call [`sort_by_file_name`](https://docs.rs/walkdir/latest/walkdir/struct.WalkDir.html#method.sort_by_file_name) on Walkdir. While Walkdir says that the order is unspecified, a little test on my system shows that the returned array is already ordered; it shouldn't be a problem to order it even for `random` sorting.

anoldguy replied:
That's even better, let Walkdir do the work for us. :raised_hands:

danyspin97 commented on daemon/src/wallpaper_info.rs:
```diff
@@ -12,6 +12,17 @@ pub struct WallpaperInfo {
+}
+
+#[derive(Debug, Copy, Clone, Default, Eq, PartialEq,Deserialize)]
+#[serde(rename = "kebab-case")]
```
Does it work without renaming? Also, do `natural` and `reverse` (in lowercase) work?

anoldguy replied:
Serde is magic to me. I freely admit to cargo-culting some of this. Happy to dig in and do some further testing. I do know that lowercase strings do not work, so I'd like to get that fixed up, too.

Rolv-Apneseth commented:
> I think that having these 3 orders will be enough for most users. I may have spent too much time working with tech people so now `Natural` sound less natural to me compared to `ascending` and `descending` (sorry for the punjoy). I am not sure if it would be better to have `random`, `ascending` and `descending` instead to be more clear with non native english speakers. But that's just a small detail that we can decide before merging.

I would agree with this and will admit: I had to search up what "natural" sort order meant ([wikipedia for anyone like me](https://en.wikipedia.org/wiki/Natural_sort_order)). The treating of numbers as a single character is pretty useful though, as some people might have images labelled as `wallpaper1...wallpaper20` and you would hope that they would go in that order. It doesn't look like `vec.sort()` does it like that though, would a crate like [human_sort](https://docs.rs/human-sort/latest/human_sort/) be required?

anoldguy commented:
> I think that having these 3 orders will be enough for most users. I may have spent too much time working with tech people so now `Natural` sound less natural to me compared to `ascending` and `descending` (sorry for the punjoy). I am not sure if it would be better to have `random`, `ascending` and `descending` instead to be more clear with non native english speakers. But that's just a small detail that we can decide before merging.

Oh, for sure. I'm not emotionally invested in `Natural` and `Reverse`, happy to switch to `ascending` and `descending`.

> Although I like the index approach that you are using, one problem that might occur is that files are added or removed from the directory; in this case the index would point to a different image than the one it should be pointing to.

Ultimately, does it matter? If we're re-reading the directory on each iteration of the loop and `wallpaper_04` gets bumped a spot by `wallpaper_03a`...do we care? I could see us caring if we keep a "wallpaper playlist" on startup, but (so far), we're very ephemeral and dynamic when it comes to directory listings, meaning that we'll slot wallpapers into their sorting order as files are added. 

I will concede that removing files will lead to us pointing to a possible recent image, or one that doesn't exist, so I should def handle that.

> One other way would be to use current_img: we search through the sorted array for current_img (getting the index), and then we take the next for Natural and the previous for Reverse. By using [binary_search](https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search), this operation should be fast enough to do it every time. What do you think of this solution?

I'm interested! More learning. I'll see what I can come up with for this.  I'm really liking the ephemeral nature of the 'walk the directory for the files each loop' process.  It makes the folder the index without keeping hardly any state.  Add a file, it shows up in your rotation with no restart needed. (Unless I'm missing something with WalkDir?)

anoldguy commented:
I've:
 
* updated the sorting options to be `ascending`, `descending`, and `random`
* updated serde to use any typical casing (`ascending`, `Ascending`, `ASCENDING`)
* updated the `WalkDir` to use `sort_by_filename`
* updated the two sorted methods to use `binary_search` on the `files` slice

danyspin97 commented on daemon/src/surface.rs:
```diff
@@ -207,7 +209,45 @@ impl Surface {
+                        if idx >= files.len() {
+                            0
+                        } else {
+                            idx + 1
```
We don't know if `idx + 1` is still inside the vector range. Since we call:

```rust
let img_path = files.into_iter().nth(index).unwrap();
```

wpaperd might panic in that case.

danyspin97 replied:
Also, we can use either `[]` operator or `get` to files without calling `into_iter`.

danyspin97 replied:
One idea is to increase the index regardless and then check afterwards.

danyspin97 commented:
> as some people might have images labelled as wallpaper1...wallpaper20 and you would hope that they would go in that order

> would a crate like [human_sort](https://docs.rs/human-sort/latest/human_sort/) be required?

That's indeed a good idea, maybe we should add `natural` sorting as well as additional sorting method. Let's get `ascending` and `descending` merged first.

> Ultimately, does it matter? If we're re-reading the directory on each iteration of the loop and wallpaper_04 gets bumped a spot by wallpaper_03a...do we care?

I feel like starting from random positions would create random situations I would prefer to avoid.

> I will concede that removing files will lead to us pointing to a possible recent image, or one that doesn't exist, so I should def handle that.

Exactly my thought. I.e. 10 images before index get deleted, the index points to images that have already been shown.

>  I'm really liking the ephemeral nature of the 'walk the directory for the files each loop' process.

I am glad you like it 😄 

> updated serde to use any typical casing (ascending, Ascending, ASCENDING)

I think we can allow just the lowercase name to have both a consistent config file and less clutter in the code. What do you think?

The PR looks good to me, once fixed the index range we can merge it :)

danyspin97 commented:
Thank you!

anoldguy commented:
@danyspin97, I can't believe I forgot `rustfmt`! 🤦‍♂️

In any case, both `surface.rs` and `wallpaper_info.rs` need it.
